### PR TITLE
Windows: Fixing straightforward issues with //test/common/network/... 

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Commit Message: Windows: Fixing straightforward issues with //test/common/network/... 
Additional Description:
- Do not use SO_REUSEADDR on Windows as it does not behave the same as Linux and other BSD socket stacks
- Disable pipe mode test on Windows, see https://github.com/envoyproxy/envoy/issues/11354
- socket_option_factory_test no longer hard codes socket option value as struct linger consists of ushorts not ints on Windows
- allocate callback handles on heap instead of stack to work around MSVC
- enable address_impl_speed_test_benchmark_test
- additionally, tag new failing tests on Windows

Risk Level: Low
Testing: Fixes Windows invocations of tests, excludes some on Windows, ran locally on Windows w/ MSVC
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
cc @wrowe

File: test/mocks/server/mocks.h [hunk 1]
```diff
@@ -691,7 +691,7 @@ class MockBootstrapExtensionFactory : public BootstrapExtensionFactory {
   MOCK_METHOD(BootstrapExtensionPtr, createBootstrapExtension,
               (const Protobuf::Message&, Configuration::ServerFactoryContext&), (override));
   MOCK_METHOD(ProtobufTypes::MessagePtr, createEmptyConfigProto, (), (override));
-  MOCK_METHOD(std::string, name, (), (const override));
+  MOCK_METHOD(std::string, name, (), (const, override));
```
Comment (ggreenway) [inline test/mocks/server/mocks.h:694]: Can you explain what's going on here? (I don't object to the change; I don't understand what is going on here, and why this broke windows).

**Comment by ggreenway** [comment]
MacOS CI failure is due to bazel version; not related to this PR.

File: test/mocks/server/mocks.h (hunk 1 above)
Comment (sunjayBhatia) [reply test/mocks/server/mocks.h:694]: MSVC doesn't accept the two specifiers without a comma, it errors with `.\test/mocks/server/mocks.h(694): error C2338: const override cannot be recognized as a valid specification modifier.`, from the google test cheatsheet, it seems a comma separated list is the correct way to specify multiple keywords: https://github.com/google/googletest/blob/master/googlemock/docs/cheat_sheet.md#mocking-a-normal-class-mockclass

Comment (ggreenway) [reply test/mocks/server/mocks.h:694]: That's really interesting. Thanks for the explanation!

**Comment by mattklein123** [comment]
Please merge master to fix CI.

/wait

**Comment by ggreenway** [comment]
Something went wrong with how you tried to merge master. You've added new, different commits with the same content as the commits on master. Please undo whatever you did and try again.

/wait

**Comment by sunjayBhatia** [comment]
> Something went wrong with how you tried to merge master. You've added new, different commits with the same content as the commits on master. Please undo whatever you did and try again.


yep, we're fixing it now

**Comment by ggreenway** [comment]
For reference, these are the commands I run to merge master, with `upstream` as this (the main) repository, starting with this PR branch checked out:

```
git fetch upstream master
git merge upstream/master
```

**Comment by wrowe** [comment]
> For reference, these are the commands I run to merge master, with `upstream` as this (the main) repository, starting with this PR branch checked out:

Thanks, we had encountered a problem with the merge and buildifier spec, and in dropping a patch ended up with a rebase rather than merge of origin/master. Fixed, and (our apologies) force-pushed to resolve.

**Comment by ggreenway** [comment]
> > For reference, these are the commands I run to merge master, with `upstream` as this (the main) repository, starting with this PR branch checked out:
> 
> Thanks, we had encountered a problem with the merge and buildifier spec, and in dropping a patch ended up with a rebase rather than merge of origin/master. Fixed, and (our apologies) force-pushed to resolve.

No problem; git is easy to mess up.

**Comment by wrowe** [comment]
@ggreenway use azp rekick to try mac over again, looks like an intermittent I/O error.

**Comment by wrowe** [comment]
And... //test/integration:hotrestart_test timing out for Linux-x64 asan.

Will need to be re-kicked, curious first whether tsa/compile_time_options both pass, first.

**Comment by mattklein123** [comment]
I've seen the hot restart tests timing out a lot recently, @jmarantz potentially due to your recent change? I can try to take a look tomorrow.

**Comment by jmarantz** [comment]
I just noticed hotrestart tests timing out but we haven't merged that change where I messing with it. That's https://github.com/envoyproxy/envoy/pull/11301 .

**Comment by jmarantz** [comment]
When testing hotrestart_test locally on an unrelated PR, it passed on 210 seconds, but I could see the machines in CI might be a bit slower than my desktop. I'm going to write a small PR to add a size-tag to this test to see if that resolves the issue.

**Comment by jmarantz** [comment]
https://github.com/envoyproxy/envoy/pull/11459

**Pull request merged**